src/app/calculator/furnaces/leakage/leakage.service.ts

```diff
@@ -0,0 +1,291 @@
+      result.grossLoss =  (result.leakageLoss / leakageLossData.availableHeat) * 100;
+      result.fuelUse = result.grossLoss * leakageLossData.hoursPerYear;
+      result.fuelCost = result.fuelUse * leakageLossData.fuelCost;
+    }
+    return result;
+  }
+
+  initDefaultEmptyInputs() {
+    let emptyBaselineData: LeakageLoss = this.initDefaultLoss(0);
+    let baselineData: Array<LeakageLoss> = [emptyBaselineData];
+    this.baselineData.next(baselineData);
+    this.modificationData.next(undefined);
+    this.energySourceType.next('Fuel');
+  }
+
+  initDefaultLoss(index: number, hoursPerYear: number = 8760) {
+    let defaultBaselineData: LeakageLoss = {
+      draftPressure: 0,
+      openingArea: 0,
+      leakageGasTemperature: 0,
+      ambientTemperature: 0,
+      coefficient: .8052,
+      specificGravity: 1.0,
+      correctionFactor: 1.0,
+      name: 'Loss #' + (index + 1),
+      energySourceType: 'Fuel',
+      fuelCost: 0,
+      hoursPerYear: hoursPerYear,    
+      availableHeat: 100,  
+    };
+
+    return defaultBaselineData;
+  }
+
+
+  initDefaultEmptyOutput() {
+     let output: LeakageLossOutput = {
+      baseline: {totalFuelUse: 0, grossLoss: 0, totalFuelCost: 0, losses: []},
+      modification: {totalFuelUse: 0, grossLoss: 0, totalFuelCost: 0, losses: []},
+      fuelSavings: 0,
+      costSavings: 0
+    };
+    this.output.next(output);
+  }
+
+  initModification() {
+    let currentBaselineData: Array<LeakageLoss> = this.baselineData.getValue();
+    let currentBaselineCopy = JSON.parse(JSON.stringify(currentBaselineData));
+    this.modificationData.next(currentBaselineCopy);
+  }
+
+  updateDataArray(data: LeakageLoss, index: number, isBaseline: boolean) {
+    let dataArray: Array<LeakageLoss>;
+    if (isBaseline) {
+      dataArray = this.baselineData.getValue();
+    } else {
+      dataArray = this.modificationData.getValue();
+    }
+    // dataArray won't exist during reset cycle w/ multiple subjects emitting
+    if (dataArray && dataArray[index]) {
+      Object.assign(dataArray[index], data);
+    }
+
+     if (isBaseline) {
+      this.baselineData.next(dataArray);
+    } else {
+      this.modificationData.next(dataArray);
+    }
+  }
+
+    
+  removeLoss(i: number) {
+    let currentBaselineData: Array<LeakageLoss> = this.baselineData.getValue();
+    currentBaselineData.splice(i, 1);
+    this.baselineData.next(currentBaselineData);
+    let currentModificationData: Array<LeakageLoss> = this.modificationData.getValue();
+    if (currentModificationData) {
+      currentModificationData.splice(i, 1);
+      this.modificationData.next(currentModificationData);
+    }
+  }
+  
+  addLoss(treasureHours: number, modificationExists: boolean) {
+    let currentBaselineData: Array<LeakageLoss> = JSON.parse(JSON.stringify(this.baselineData.getValue()));
+    let index = currentBaselineData.length;
+    let hoursPerYear = treasureHours? treasureHours : currentBaselineData[index - 1].hoursPerYear;
```

Comment:
index-1 breaks when no baseline data. `index-1` always big red flag

Reply:
Suggestion.. 

instead of continually accessing the object in the array using `baselineData[index-1]`

add a third optional parameter to `initDefaultLoss` that is a LeakageLoss.
pass `baselineData[0]` 
if it exists, use it to set the `fuelCost` and `availableHeat` of the new loss